### PR TITLE
Add x-bit ints

### PR DESCRIPTION
**Pull request by abhra2020-smart**
Add customizability to how much bit ints there are

### CODE REVIEW

**Comment by NeeEoo** [comment]
The interpreter should be able to receive the bit size from a parameter on the interpreter function. So that you can pass a value from the compiler with for example `--interpreter-bits 16`

**Comment by NeeEoo** [comment]
I think you forgot to add the expression to the `-` instruction

File: Interpreter.py [hunk 1]
```diff
@@ -40,7 +41,7 @@ def brainfuck(program):
             data_pointer -= 1
         elif command == '+':
             data[data_pointer] = (data.get(data_pointer, 0) + 1)
-            if data[data_pointer] == 256:
+            if data[data_pointer] == (int(255 / 8) * bits) + 1:
```
Comment (elikaski) [inline Interpreter.py:44]: 2**bits

**Comment by elikaski** [comment]
Please make the interpreter receive an optional flag indicating the cell bit size :)

**Review by NeeEoo — CHANGES_REQUESTED** [review]
I forgot to submit these reviews. They are from 14 hours ago.

**Comment by NeeEoo** [comment]
When this gets merged (after the changes) i will make some library functions that are related to this and are very much needed.

File: Interpreter.py [hunk 2]
```diff
@@ -64,13 +65,20 @@ def brainfuck(program):
...
     if len(sys.argv) != 2:
-        print("Usage: '%s' <path_to_brainfuck_code_file>" % sys.argv[0])
+        print("Usage: '%s' <path_to_brainfuck_code_file>\nUsage: '%s' <path_to_brainfuck_file> -bits <x-bit ints>\nUsage: '%s' <path_to_brainfuck_file> --interpreter-bits <x-bit ints>" % (sys.argv[0], sys.argv[0]))
         exit(0)
 
-    fpath = sys.argv[1]
-    with open(fpath, "rt") as f:
-        code = f.read()
+    parser = argparse.ArgumentParser()
+    parser.add_argument("filepath", metavar="fp")
```
Comment (NeeEoo) [inline Interpreter.py:75]: Change this to `parser.add_argument("filepath", metavar="input_file", nargs=1, help="Path to brainfuck code file")`

File: Interpreter.py [hunk 3]
```diff
@@ -64,13 +65,20 @@ def brainfuck(program):
...
+    parser.add_argument("-bits", "--interpreter-bits", metavar="bits", nargs=1)
```
Comment (NeeEoo) [inline Interpreter.py:76]: Change this to `parser.add_argument("--bits", "-b", nargs=1, default=8, help="Amount of bits each cell uses")`

File: Interpreter.py [hunk 4]
```diff
@@ -64,13 +65,20 @@ def brainfuck(program):
...
+
+    args = parser.parse_args()
+    bits = args.bits  #  TODO: Doesn't work for me
+    f = open(args.fp, 'r')
```
Comment (NeeEoo) [inline Interpreter.py:80]:
```py
with open(args.fp, "r") as f:
    code = f.read()

brainfuck(code, bits)
```

File: Interpreter.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env python3
 
 import sys
+import argparse
 
 
 def create_jumps_dictionary(program):
```
Comment (NeeEoo) [inline Interpreter.py:7]: Add `,  bits=8`

**Review by NeeEoo — CHANGES_REQUESTED** [review]

File: Interpreter.py (hunk 3 above)
Comment (abhra2020-smart) [reply Interpreter.py:76]: Thanks! I don't even know how to use argparse lol.

**Comment by NeeEoo** [comment]
You should pass the bits variable to the brainfuck function. And in the definition https://github.com/elikaski/BF-it/blob/735f276cef40e0e567a473cf8871b11a4ca70180/Interpreter.py#L27
You should add `, bits=8` so you can call this function from another file. Since you can run the compiled code from the compiler

File: Interpreter.py [hunk 6]
```diff
@@ -40,7 +41,7 @@ def brainfuck(program):
             data_pointer -= 1
         elif command == '+':
             data[data_pointer] = (data.get(data_pointer, 0) + 1)
-            if data[data_pointer] == 256:
+            if data[data_pointer] == (int(255 / 8) * int(bits)) + 1:
```
Comment (elikaski) [inline Interpreter.py:44]: Please change this line to `2 ** bits`, it is much clearer than this calculation :)

File: Interpreter.py [hunk 7]
```diff
@@ -40,7 +41,7 @@ def brainfuck(program):
...
                 data[data_pointer] = 0
         elif command == '-':
```
Comment (elikaski) [inline Interpreter.py:46]: Please use the correct amount of bits here as well. Currently it is hard-coded to 255

File: Interpreter.py [hunk 8]
```diff
@@ -64,13 +65,23 @@ def brainfuck(program):
     if data_pointer != 0:
         print("WARNING (interpreter) - at the end of the execution the data pointer is %s instead of 0 (possibly a compiler issue)" % str(data_pointer))
 
+
 if __name__ == '__main__':
-    if len(sys.argv) != 2:
-        print("Usage: '%s' <path_to_brainfuck_code_file>" % sys.argv[0])
+    if len(sys.argv) < 2:
```
Comment (elikaski) [inline Interpreter.py:70]: This check can be removed, as the argparse library deals it for us :)

File: Interpreter.py [hunk 9]
```diff
@@ -64,13 +65,23 @@ def brainfuck(program):
...
+        print("Usage: '%s' <path_to_brainfuck_code_file>\nUsage: '%s' <path_to_brainfuck_file> -b <x-bit ints>\nUsage: '%s' <path_to_brainfuck_file> --bits <x-bit ints>" % (sys.argv[0], sys.argv[0], sys.argv[0]))
         exit(0)
 
-    fpath = sys.argv[1]
-    with open(fpath, "rt") as f:
-        code = f.read()
+    parser = argparse.ArgumentParser()
+    parser.add_argument("filepath")
+    parser.add_argument("--bits", "-b", "--interpreter-bits", nargs=1, default=8, help="Amount of bits each cell uses")
```
Comment (elikaski) [inline Interpreter.py:76]: Please add `type=int`

Comment (elikaski) [reply Interpreter.py:76]: and remove `nargs=1` so it will not be a list

File: Interpreter.py [hunk 10]
```diff
@@ -64,13 +65,23 @@ def brainfuck(program):
...
+
+    args = parser.parse_args()
+    try:
```
Comment (elikaski) [inline Interpreter.py:79]: This check is not needed, because argparse deals with it for us :)

File: Interpreter.py [hunk 11]
```diff
@@ -64,13 +65,23 @@ def brainfuck(program):
...
+        bits = args.bits[0]
+    except:
+        bits = 8
+    f = open(args.filepath, 'r')
+    code = f.read()
+    f.close()
 
-    brainfuck(code)
+    brainfuck(code, bits)
```
Comment (elikaski) [inline Interpreter.py:87]: brainfuck(code, args.bits)

**Review by elikaski — CHANGES_REQUESTED** [review]
Hey, thanks for your commit :)
I've made some comments, mostly about unnecessary things that argparse library implements for us
Please make the changes and I'll merge :)

File: Interpreter.py (hunk 6 above)
Comment (abhra2020-smart) [reply Interpreter.py]: Forgot how bits work lol.
Uninstalling Windows out of rage...
Uninstalled Windows.
Throw computer out of window.
Threw computer out of window.
jk not really

File: Interpreter.py (hunk 8 above)
Comment (abhra2020-smart) [reply Interpreter.py]: Life was never easier :)

**Comment by abhra2020-smart** [comment]
Made changes.

**Comment by abhra2020-smart** [comment]
WTH WHY DID I CLOSE IT?!?!?!

**Comment by elikaski** [comment]
Excellent! Thank you :)
You missed 2 things that I mentioned. I tagged you in them so you will see
They are:
1. Adding `nargs=1` to the -b argument
2. Adding the bits calculation to the "-" instruction, similar to the "+" instruction
Currently the value "255" is hard-coded, it should be `2 ** bits - 1`

Please add these and I'll merge :)

**Comment by elikaski** [comment]
> Excellent! Thank you :)
> You missed 2 things that I mentioned. I tagged you in them so you will see
> They are:
> 
> 1. Adding `nargs=1` to the -b argument
> 2. Adding the bits calculation to the "-" instruction, similar to the "+" instruction
>    Currently the value "255" is hard-coded, it should be `2 ** bits - 1`
> 
> Please add these and I'll merge :)

(for some reason I can't tag you, so I kept these conversations as "unresolved")

**Comment by abhra2020-smart** [comment]
Hopefully the final commit has been made

**Comment by NeeEoo** [comment]
I think that everything looks good. Only thing is the file opening. The old version used `with open(...) as f:`

**Comment by NeeEoo** [comment]
The f.close() is unnecessary since the with statement does that automatically

**Comment by elikaski** [comment]
Merged, thank you :)

**Pull request merged**